Samples/Widgets/SingleWidget/cpp-win32-packaged/SampleWidgetProviderApp/WidgetProvider.cpp

```diff
@@ -0,0 +1,205 @@
+	auto widgetContext = actionInvokedArgs.WidgetContext();
+	auto widgetId = widgetContext.Id();
+	if (auto&& widget = FindRunningWidget(widgetId))
+	{
+		if (verb == L"inc")
+		{
+			widget->m_clickCount++;
+			// Generate template/data you want to send back
+			// The template has not changed so it does not neeed to be updated
+			auto widgetData = GetDataForWidget(widgetId);
+
+			// Build update options and update the widget
+			winrt::WidgetUpdateRequestOptions updateOptions{ widgetId };
+			updateOptions.Data(widgetData);
+			updateOptions.CustomState(GetStateForWidget(widgetId));
+
+			winrt::WidgetManager::GetDefault().UpdateWidget(updateOptions);
+			return;
+		}
+		throw winrt::hresult_invalid_argument{ L"OnActionInvoked invoked for widget with id: " + widgetId + L" and unknown verb: " + verb};
+	}
+	throw winrt::hresult_invalid_argument{ L"OnActionInvoked invoked for unknown widget with id: " + widgetId };
+}
+
+// Handle the WidgetContextChanged call. This function is called when the context a widget
+// has changed. Currently it only signals that the user has changed the size of the widget.
+// There are 2 ways to respond to this event:
+// 1) Call UpdateWidget() with the new data/template to fit the new requested size.
+// 2) If previously sent data/template accounts for various sizes based on $host.widgetSize - you can use this event solely for telemtry.
+void WidgetProvider::OnWidgetContextChanged(winrt::WidgetContextChangedArgs contextChangedArgs)
+{
+	auto widgetContext = contextChangedArgs.WidgetContext();
+	auto widgetId = widgetContext.Id();
+	if (auto&& widget = FindRunningWidget(widgetId))
+	{
+		// Optionally: if the data/template for the new context is different
+		// from the previously sent data/template - send an update.
+	}
+	throw winrt::hresult_invalid_argument{ L"WidgetContextChanged invoked for unknown widget with id: " + widgetId };
+}
+
+// Handle the Activate call. This function is called when widgets host starts listening
+// to the widget updates. It generally happens when the widget becomes visible and the updates
+// will be promptly displayed to the user. It's recommended to start sending updates from this moment
+// until Deactivate function was called.
+void WidgetProvider::Activate(winrt::Microsoft::Windows::Widgets::Providers::WidgetContext widgetContext)
+{
+	auto widgetId = widgetContext.Id();
+	if (auto&& widget = FindRunningWidget(widgetId))
+	{
+		widget->m_isActivated = true;
+
+		winrt::WidgetUpdateRequestOptions updateOptions{ widgetId };
+		updateOptions.Template(GetTemplateForWidget());
+		updateOptions.Data(GetDataForWidget(widgetId));
+		// You can store some custom state in the widget service that you will be able to query at any time.
+		updateOptions.CustomState(GetStateForWidget(widgetId));
+		// Update the widget
+		winrt::WidgetManager::GetDefault().UpdateWidget(updateOptions);
+
+	}
+	throw winrt::hresult_invalid_argument{ L"Trying to deactivate unknown widget with id: " + widgetId };
```

Comment:
Defensive coding - is the manager object on the other side going to intercept this error and "do something" with it? Like stop sending requests to this provider? Maybe this "activate for invalid id" should generate a specific hresult that the manager can observe and know the provider has lost the given widget?

Reply:
Nope, but it might be a good idea for the `Manager` to react to exceptions.
I've added this exception so that the developer can quickly see what went wrong in this function call. If a call came for a widget that provider is not aware of - it will need to ask the `WidgetManager` for all the running widgets that this provider is supposedly serving (same thing that's done in provider constructor). May be it's a good idea to add explaining comment?